ttnn/cpp/ttnn/operations/moreh/moreh_adam/device/moreh_adam_device_operation.cpp

```diff
@@ -62,57 +62,49 @@ void MorehAdamOperation::validate_on_program_cache_hit(
     validate_inputs(operation_attributes, tensor_args);
 };
 
-MorehAdamOperation::shape_return_value_t MorehAdamOperation::compute_output_shapes(
+MorehAdamOperation::spec_return_value_t MorehAdamOperation::compute_output_specs(
     const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
-    auto input_tensor_shape = tensor_args.param_in.get_shape();
+    auto output_shape = tensor_args.param_in.get_logical_shape();
+    auto dtype = tensor_args.param_in.get_dtype();
 
-    return {
-        input_tensor_shape,
-        input_tensor_shape,
-        input_tensor_shape,
-        input_tensor_shape,
-    };
-};
+    std::vector<std::optional<TensorSpec>> ret;
+    TensorSpec out_spec(
+        output_shape, TensorLayout(dtype, PageConfig(Layout::TILE), operation_attributes.memory_config));
+    for (int idx = 0; idx < 3; idx++) {
+        if (tensor_args.output_tensors.at(idx).has_value()) {
+            ret.push_back(tensor_args.output_tensors.at(idx)->get_tensor_spec());
+        } else {
+            ret.push_back(out_spec);
+        }
+    }
+    if (tensor_args.output_tensors.at(3).has_value()) {
+        ret.push_back(tensor_args.output_tensors.at(3)->get_tensor_spec());
+    } else if (operation_attributes.amsgrad) {
+        ret.push_back(out_spec);
+    } else {
+        ret.push_back(std::nullopt);
+    }
```

Comment:
@sminakov-tt , I am confused, whats going on here?

Reply:
it was not accurate before?

Reply:
Yes, previously it was inconsistent with  create_output_tensors